[shazxrin] iP

shazxrin opened the pull request:
# Duke

> "Habits are the compound interest of self-improvement." - James Clear ([source](https://www.goodreads.com/work/quotes/62221762-atomic-habits-an-easy-proven-way-to-build-good-habits-break-bad-one))

Duke is a task manager that aids you in achieving your _goals_. It's
- text-based
- clear and concise
- ~~fast~~ **super fast**
- **_free for everyone_** :smile:

## Get Started

1. Download [here](https://github.com/shazxrin/ip/releases/tag/A-Jar)
2. Run it via `java -jar duke.jar`

## Features
- [x] Keep track of todos
- [x] Keep track of events
- [x] Keep track of deadlines
- [ ] Keep track of reminders (coming soon)

## Integration

Want to integrate Duke into your own application? Just add Duke in your application `main` method:
```java
public class YourApplication {
    public static void main(String[] args) {
        // Your application code.
        new Duke("<your save file directory>", "<your save file name>").run()
    }
}
```

Commit:
Add Gradle support

shazxrin committed:
Base skeleton for chatbot

shazxrin committed:
Add task and list tasks feature

shazxrin committed:
Mark and unmark tasks feature

shazxrin committed:
Add ToDos, Events, Deadlines feature

shazxrin committed:
Clean up command handling

shazxrin committed:
Add test input and expected test output

shazxrin committed:
Add DukeException class

shazxrin committed:
Error handling for incorrect and unknown command

shazxrin committed:
Error handling for mark and unmark commands

shazxrin committed:
Delete task feature

shazxrin committed:
Add tasks persistence

shazxrin committed:
Add date & time for deadline and event

shazxrin committed:
Update tests with actual date time

shazxrin committed:
Merge branch 'branch-Level-7'

shazxrin committed:
Merge branch 'branch-Level-8'

shazxrin committed:
Reorganize program into separate OOP classes

shazxrin committed:
Move classes into their respective packages

shazxrin committed:
Merge branch 'add-gradle-support'

shazxrin committed:
Setup gradle for tests and run

shazxrin committed:
Setup JUnit tests

shazxrin committed:
Add JavaDoc comments

shazxrin committed:
Tweak code to comply to coding standard

shazxrin committed:
Add find tasks command

shazxrin committed:
Merge branch 'branch-A-JavaDoc'

shazxrin committed:
Merge branch 'branch-A-CodingStandard'

shazxrin committed:
Merge branch 'branch-Level-9'

shazxrin committed:
Add checkstyle and align codebase accordingly

Skeltons commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,56 @@
+
+import java.util.Scanner;
+
+public class Duke {
```
Shouldn't there be a Javadoc comment for public classes?

shazxrin replied:
Good catch! Thank you 😄

Skeltons commented on src/main/java/duke/storage/FileStorage.java:
```diff
@@ -0,0 +1,122 @@
+        try {
+            FileWriter fileWriter = new FileWriter(String.format("%s/%s", directoryPath, fileName));
+            for (Task task : tasks) {
+                switch (task.getTaskType()) {
+                case TODO:
+                    ToDo todo = (ToDo) task;
+                    fileWriter.write(String.format("T;%d;%s\n", todo.isDone() ? 1 : 0, todo.getDescription()));
+
+                    break;
+                case DEADLINE:
+                    Deadline deadline = (Deadline) task;
+                    fileWriter.write(String.format("D;%d;%s;%s\n", deadline.isDone() ? 1 : 0, deadline.getDescription(),
+                            deadline.getBy().format(DateTimeUtils.DATE_TIME_FORMAT_INPUT)));
+
+                    break;
+                case EVENT:
+                    Event event = (Event) task;
+                    fileWriter.write(String.format("E;%d;%s;%s;%s\n", event.isDone() ? 1 : 0, event.getDescription(),
+                            event.getFrom().format(DateTimeUtils.DATE_TIME_FORMAT_INPUT),
+                            event.getTo().format(DateTimeUtils.DATE_TIME_FORMAT_INPUT)));
```
Shouldn't there be no indentations between switch and case?

shazxrin replied:
Yes, there should not be indentations. The case blocks are not indented inside the switch block.

Skeltons reviewed (commented):
Overall code LGTM! I like how you recognize that the 3 command `todo`, `deadline` and `event` adds task to the task list and created an abstract superclass `AddTaskCommand`  to encapsulate them.

tingyic commented on src/main/java/duke/command/AddTaskCommand.java:
```diff
@@ -0,0 +1,32 @@
+package duke.command;
```
I like how you separated different functions into different packages. :)

tingyic commented on src/main/java/duke/command/CommandParser.java:
```diff
@@ -0,0 +1,93 @@
+
+    private boolean hasUserQuit = false;
+
+    public CommandParser() {
```
That's a neat way to parse in different commands 👍

tingyic commented on src/main/java/duke/command/DeadlineCommand.java:
```diff
@@ -0,0 +1,50 @@
+import java.time.LocalDateTime;
+import java.time.format.DateTimeParseException;
+
+public class DeadlineCommand extends AddTaskCommand {
```
Great use of OOP here! Similar for DeleteCommand, EventCommand, and ToDoCommand.

tingyic commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,38 @@
+
+public abstract class Task {
+    private String description;
+    private TaskType taskType;
```
nitpicking here a bit, but don't you think typeOfTask would sound better?

shazxrin replied:
I agree! 😅

shazxrin committed:
Add graphical user interface

shazxrin committed:
Use variable arguments in commands

shazxrin committed:
Merge branch 'branch-A-CheckStyle'

shazxrin committed:
Merge branch 'branch-Level-10'

shazxrin committed:
Merge branch 'branch-A-Varargs'

shazxrin committed:
Fix code to align with checkstyle

shazxrin committed:
Rename variable for code quality

shazxrin committed:
Add assert statements

Assert statements added to all command main logic to ensure that the arguments passed are of the correct length.

shazxrin committed:
Merge pull request #1 from shazxrin/branch-A-Assertions

Add assert statements

shazxrin committed:
Extract methods from long methods

Logic in loading and saving tasks into and from file has been extracted into methods. This is to align with the code quality standard: avoid long methods.

By doing so, the code is more readable and easier to follow.

shazxrin committed:
Merge pull request #2 from shazxrin/branch-A-CodeQuality

Extract methods from long methods

shazxrin committed:
Convert listing logic using streams

Using streams makes the code more intuitive and easier to follow as it reduces the number of lines required.

shazxrin committed:
Merge pull request #3 from shazxrin/branch-A-Streams

Convert listing logic using streams

shazxrin committed:
Add places into tasks

Add places into the three types of task to indicate where the task should be done.

shazxrin committed:
Merge branch 'branch-D-Places'

shazxrin committed:
Update GUI

shazxrin committed:
Merge branch 'branch-A-BetterGUI'

shazxrin committed:
Improve text messages for personality

shazxrin committed:
Merge branch 'branch-A-Personality'

shazxrin committed:
Add error handling for empty list

shazxrin committed:
Merge branch 'branch-A-MoreErrorHandling'

shazxrin committed:
Add tests for list and find commands

shazxrin committed:
Merge branch 'branch-A-MoreTesting'

shazxrin committed:
Replace copyrighted images

shazxrin committed:
Add user guide and UI screenshot

shazxrin committed:
Fix header comments according to standard

shazxrin committed:
Update user guide README

shazxrin committed:
Fix UI based on peer feedback
